servicetalk-concurrent-api/src/test/java/io/servicetalk/concurrent/api/ScanWithPublisherTest.java

```diff
@@ -201,68 +307,220 @@ public boolean mapTerminal() {
+            @Override
+            public boolean mapTerminal() {
+                return true;
+            }
+
+            @Override
+            public void afterFinally() {
+                finalizations.incrementAndGet();
+            }
+        })).subscribe(subscriber);
+
+        Subscription s = subscriber.awaitSubscription();
+        s.request(1);
+        processor.onNext(1);
+        processor.onComplete();
+        s.cancel();
+
+        assertThat(finalizations.get(), is(1));
     }
 
-    @Test
-    public void onErrorThrowsHandled() {
-        terminalThrowsHandled(false);
+    @ParameterizedTest(name = "{displayName} [{index}] {arguments}")
+    @ValueSource(booleans = {true, false})
+    public void scanWithFinalizationOnCancelDifferentThreads(final boolean interleaveCancellation)
+            throws InterruptedException {
+        final AtomicInteger finalizations = new AtomicInteger(0);
+        PublisherSource.Processor<Integer, Integer> processor = newPublisherProcessor();
+
+        final CountDownLatch checkpoint = new CountDownLatch(1);
+        final CountDownLatch requested = new CountDownLatch(1);
+        final CountDownLatch resume = new CountDownLatch(1);
+
+        final CountDownLatch nextDelivered = new CountDownLatch(1);
+        final CountDownLatch nextDeliveredResume = new CountDownLatch(1);
+
+        final PublisherSource<Integer> source = toSource(scanWithOperator(fromSource(processor), true,
+                new ScanWithLifetimeMapper<Integer, Integer>() {
+            @Override
+            public Integer mapOnNext(@Nullable final Integer next) {
+                return next;
+            }
+
+            @Override
+            public Integer mapOnError(final Throwable cause) throws Throwable {
+                throw cause;
+            }
+
+            @Override
+            public Integer mapOnComplete() {
+                return 5;
+            }
+
+            @Override
+            public boolean mapTerminal() {
+                if (interleaveCancellation) {
+                    checkpoint.countDown();
+                    try {
+                        resume.await();
+                    } catch (InterruptedException e) {
+                        e.printStackTrace();
+                    }
+                }
+                return true;
+            }
+
+            @Override
+            public void afterFinally() {
+                finalizations.incrementAndGet();
+            }
+        }));
+
+        Thread sub = new Thread(() -> {
+            TestPublisherSubscriber<Integer> subscriber = new TestPublisherSubscriber<>();
+            source.subscribe(subscriber);
+            Subscription s = subscriber.awaitSubscription();
+
+            s.request(2);
+            requested.countDown();
+
+            if (interleaveCancellation) {
+                try {
+                    checkpoint.await();
+                } catch (InterruptedException e) {
+                    e.printStackTrace();
```

Comment:
a common practice for tests with other threads where tasks may throw is as follows:
```java
ExecutorService e = ...;
try {
  Future<Void> f = e.submit(() -> {
    try {
      methodThatMayThrow();
     } catch (Exception e) {
        throw new AssertionError(e);
     }
    return null;
   });

   f.get();
} finally {
  e.shutdown();
}
```

this provides visibility to the test thread (and therefore the junit report) if an exception occurs in another thread, and also allows you to wait for the task to complete if required.